accel/sel4/sel4-all.c

```diff
@@ -189,76 +192,114 @@ static void sel4_change_state_handler(void *opaque, bool running, RunState state
+        dev->config_write(dev, addr, val, len);
     }
+}
+
+static inline int handle_mmio(SeL4State *s, rpcmsg_t *req)
+{
+    seL4_Word dir = BIT_FIELD_GET(req->mr0, RPC_MR0_MMIO_DIRECTION);
+    seL4_Word as = BIT_FIELD_GET(req->mr0, RPC_MR0_MMIO_ADDR_SPACE);
+    seL4_Word len = BIT_FIELD_GET(req->mr0, RPC_MR0_MMIO_LENGTH);
+    seL4_Word slot = BIT_FIELD_GET(req->mr0, RPC_MR0_MMIO_SLOT);
+    seL4_Word addr = req->mr1;
+    seL4_Word data = req->mr2;
+
+    qemu_mutex_lock_iothread();
+
+    if (as == AS_GLOBAL) {
+        sel4_mmio_do_io(dir, addr, &data, len);
+    } else {
+        sel4_pci_do_io(as, dir, addr, &data, len);
+    }
+
     qemu_mutex_unlock_iothread();
+
+    return driver_ack_mmio_finish(&s->rpc, slot, data) ? 0 : -1;
 }
 
-static inline void handle_ioreq(SeL4State *s)
+static int handle_msg(SeL4State *s, rpcmsg_t *msg)
 {
-    struct sel4_ioreq *ioreq;
-    int slot;
-
-    for (slot = 0; slot < SEL4_MAX_IOREQS; slot++) {
-        ioreq = &s->ioreq_buffer->request_slots[slot];
-        if (qatomic_load_acquire(&ioreq->state) == SEL4_IOREQ_STATE_PROCESSING) {
-            if (ioreq->addr_space == AS_GLOBAL) {
-                sel4_mmio_do_io(ioreq);
-            } else {
-                sel4_pci_do_io(ioreq);
-            }
+    unsigned int op = BIT_FIELD_GET(msg->mr0, RPC_MR0_OP);
 
-            sel4_vm_ioctl(s, SEL4_NOTIFY_IO_HANDLED, slot);
-        }
+    switch (op) {
+    case RPC_MR0_OP_MMIO:
+        return handle_mmio(s, msg);
+    default:
+        /* no idea what to do */
+        return -1;
     }
+
+    return 0;
+}
+
+static inline int sel4_vm_wait_for_io(SeL4State *s)
+{
+    return sel4_vm_ioctl(s, SEL4_WAIT_IO, 0);
 }
 
 static void *do_sel4_virtio(void *opaque)
 {
     SeL4State *s = opaque;
+    rpcmsg_t *msg;
     int rc;
 
     for (;;) {
-        rc = sel4_vm_ioctl(s, SEL4_WAIT_IO, 0);
+        sel4_rpc_doorbell(&s->rpc);
+
+        rc = sel4_vm_wait_for_io(s);
         if (rc)
             continue;
 
-        handle_ioreq(s);
+        while ((msg = rpcmsg_queue_head(s->rpc.rx_queue)) != NULL) {
+            seL4_Word state = BIT_FIELD_GET(msg->mr0, RPC_MR0_STATE);
+
+            if (state == RPC_MR0_STATE_RESERVED) {
+                /* VMM side is still crafting the message, try again later */
+                break;
+            } else if (state == RPC_MR0_STATE_PENDING) {
+                /* kernel has not examined this yet, try again later */
+                break;
+            } else if (state != RPC_MR0_STATE_PROCESSING) {
+                /* kernel is supposed to discard completed ioeventfds */
+                error_report("corrupted msgqueue");
+                exit(1);
+            }
+
+            rc = handle_msg(s, msg);
+            if (rc) {
+                error_report("handle_msg() failed (%d)", rc);
+                exit(1);
+            }
+
+            msg->mr0 = BIT_FIELD_SET(msg->mr0, RPC_MR0_STATE, RPC_MR0_STATE_COMPLETE);
+
+            /* smp_wmb()? */
```

Comment:
Yes, a write barrier would be appropriate here because it ensures mr0 is flushed before andvance